TodoApp/test/TodoApp.Api.Tests/Controllers/ItemsControllerTests.cs

```diff
@@ -78,24 +91,59 @@ public void SetUp()
             Assert.That(item, Is.EqualTo(_items[0]).UsingItemEqualityComparer());
         }
 
+        [Test]
+        public async Task GetAsync_EmptyId_ReturnsBadRequest()
+        {
+            _controller.ModelState.Clear();
+
+            var response = await GetResultFromAction(controller => controller.GetAsync(Guid.Empty));
+
+            Assert.That(response.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest));
+        }
+
+        [Test]
+        public async Task GetAsync_NonExistentId_ReturnsNotFound()
+        {
+            _getItemByIdService.GetItemByIdAsync(Guid0).Returns(new RetrievedItem {WasFound = false});
+
+            var (contentResult, item) = await GetResultFromAction<Item>(controller => controller.GetAsync(Guid0));
+
+            Assert.That(contentResult, Is.Not.Null);
+            Assert.That(item, Is.EqualTo(null));
+            Assert.That(contentResult.StatusCode, Is.EqualTo(HttpStatusCode.NotFound));
+        }
+
         [Test]
         public async Task PostAsync_NewItem_SetsLocationHeaderReturnsItemToPost()
         {
-            _service.AddItemAsync(ItemToPost).ReturnsForAnyArgs(ItemToPost);
-            _helper.GetUriLocation(Arg.Any<Guid>()).Returns(Uri);
+            _addItemService.AddItemAsync(Arg.Is<Item>(value => value.Text == ItemToPost.Text)).Returns(ItemToPost);
+            _helper.GetUriLocation(ItemToPost.Id).Returns(Uri);
 
-            var (createdResult, item) = await GetResultFromAction<Item>(controller => controller.PostAsync(ItemToPost));
+            var (createdResult, item) = await GetResultFromAction<Item>(controller => controller.PostAsync(new Item{Text = "itemToPost"}));
             var location = createdResult.Headers.Location.ToString();
 
             Assert.That(createdResult.StatusCode, Is.EqualTo(HttpStatusCode.Created));
             Assert.That(item, Is.EqualTo(ItemToPost).UsingItemEqualityComparer());
             Assert.That(location, Is.EqualTo(UriString));
         }
 
+        [TestCaseSource(nameof(InvalidItems))]
+        public async Task PostAsync_InvalidItem_ReturnsBadRequest(Item item)
+        {
+            _controller.ModelState.Clear();
+
+            var response = await GetResultFromAction(controller => controller.PostAsync(item));
+
+            Assert.That(response.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest));
+        }
+
         [Test]
         public async Task PutAsync_ItemToUpdateWithExistingId_ReturnsUpdatedItem()
         {
-           _repository.UpdateAsync(_items[0]).ReturnsForAnyArgs(_items[1]);
+           _repository.UpdateAsync(Arg.Is<Item>(value 
+               => value.Id == _items[1].Id 
```

Comment:
Also, this is a bit complex logic here and in the `post` test as well. Please, move it to a named private method (it will be a lamda actually) - you will improve readability of your test…